File: rust/src/solutions/s0004.rs
use crate::utils;
use proconio::input;
use std::time::Instant;

/// Solution for Project Euler problem 4
///
/// Copyright michimani All rights reserved.
///
/// https://projecteuler.net/problem=4
pub fn solve() {
    let org = 3;
    println!("(original: {})", org);
    input! {
      n: u32,
    }

    let start = Instant::now();

    let ten: usize = 10;
    let mut answer = 0;

    for a in (ten.pow(n - 1))..(ten.pow(n) + 1) {
        for b in (ten.pow(n - 1))..(ten.pow(n) + 1) {
            let times = a * b;
            let is_p = utils::sequence::is_palindromic(&times.to_string());
            if is_p && times > answer {
                answer = times;
            }
        }
    }

    println!("answer is {}", answer);

    let end = start.elapsed();
    println!(
        "\nIt took {}.{:03} seconds.",
        end.as_secs(),
        end.subsec_nanos() / 1_000_000
    );
}

File: python/solution/s0004.py
#
# Solution for Project Euler problem 4
# Copyright michimani All rights reserved.
#
# https://projecteuler.net/problem=4
#

DIGIT = 3


def solve():
    palindrome_product_list = list()
    for a in range(10 ** (DIGIT - 1), 10**DIGIT):
        for b in range(10 ** (DIGIT - 1), 10**DIGIT):
            product = a * b
            if is_palindrome(product) is True:
                palindrome_product_list.append(product)

    return max(palindrome_product_list)


def is_palindrome(number):
    return str(number) == "".join(reversed(list(str(number))))


if __name__ == "__main__":
    print(solve())
